src/dotnet/commands/dotnet-build/DotnetProjectBuilder.cs

```diff
@@ -0,0 +1,243 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Microsoft.DotNet.Cli;
+using Microsoft.DotNet.Cli.Compiler.Common;
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.DotNet.ProjectModel;
+using Microsoft.DotNet.Tools.Compiler;
+using Microsoft.Extensions.PlatformAbstractions;
+
+namespace Microsoft.DotNet.Tools.Build
+{
+    class DotNetProjectBuilder : ProjectBuilder
+    {
+        private readonly BuilderCommandApp _args;
+        private readonly IncrementalPreconditionManager _preconditionManager;
+        private readonly CompilerIOManager _compilerIOManager;
+        private readonly ScriptRunner _scriptRunner;
+        private readonly DotNetCommandFactory _commandFactory;
+
+        public DotNetProjectBuilder(BuilderCommandApp args) : base(args.ShouldSkipDependencies)
+        {
+            _args = (BuilderCommandApp)args.ShallowCopy();
+            _preconditionManager = new IncrementalPreconditionManager(
```

Comment:
_preconditionManager should be used to compute preconditions and then have incrementality switched on / off based on its results

I'd make two visitors on the graph: a visitor that computes incremental preconditions (aggregates preconditions from all nodes) and another visitor that builds (mostly what's in ProjectBuilder) the graph. The result of the first visitor (whether the entire graph is safe or not for incrementality) affects the way the second visitor is setup (incremental on / off)